src/packages/utils/tests/logger.test.ts

```diff
@@ -0,0 +1,182 @@
+
+const getFileDescriptor = (slug: string) => {
+  const path = `./tests/test-${slug}.log`;
+  return {
+    path,
+    descriptor: openSync(path, "a")
+  };
+};
+
+describe("createLogger()", () => {
+  const timestampRegex = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z$/;
+
+  const createbaseLoggerger = () => {
+    const calls: any[][] = [];
+    return {
+      baseLogger: {
+        log: (message, ...params) => {
+          calls.push([message, ...params]);
+        }
+      },
+      calls
+    };
+  };
+
+  const message = "test message";
+
+  describe("log()", () => {
+    it("should create a baseLogger() logger", () => {
+      const { baseLogger, calls } = createbaseLoggerger();
+      const { log } = createLogger({ baseLogger });
+
+      log(message);
+
+      assert.strictEqual(
+        calls.length,
+        1,
+        "baseLogger() was called unexpected number of times."
+      );
+
+      const baseLoggerArgs = calls[0];
+
+      assert.deepStrictEqual(
+        baseLoggerArgs,
+        [message],
+        "baseLogger() called with unexpected arguments."
+      );
+    });
+
+    it("should still call baseLogger() when a file is specified", async () => {
+      const { descriptor, path } = getFileDescriptor("write-to-console");
+      const { baseLogger, calls } = createbaseLoggerger();
+
+      const { log, getCompletionHandle } = createLogger({
+        file: descriptor,
+        baseLogger
+      });
+
+      try {
+        log(message);
+        await getCompletionHandle();
+      } finally {
+        closeSync(descriptor);
+        await unlink(path);
+      }
+
+      assert.strictEqual(
+        calls.length,
+        1,
+        "baseLogger() was called unexpected number of times."
+      );
+
+      const args = calls[0];
+
+      assert.deepStrictEqual(
+        args,
+        [message],
+        "baseLogger() called with unexpected arguments."
+      );
+    });
+
+    it("should write to the file provided", async () => {
+      const { descriptor, path } = getFileDescriptor("write-to-file-provided");
+      const { baseLogger } = createbaseLoggerger();
+
+      const { log, getCompletionHandle } = createLogger({
+        file: descriptor,
+        baseLogger
+      });
+
+      let fileContents: string;
+      try {
+        log(`${message} 0`);
+        log(`${message} 1`);
+        log(`${message} 2`);
+        await getCompletionHandle();
+
+        fileContents = await readFile(path, "utf8");
+      } finally {
+        closeSync(descriptor);
+        await unlink(path);
+      }
+
+      const logLines = fileContents.split("\n");
+
+      // 4, because there's a \n at the end of each line
+      assert.strictEqual(logLines.length, 4);
+      assert.strictEqual(logLines[3], "");
+
+      logLines.slice(0, 3).forEach((logLine, lineNumber) => {
+        const timestampPart = logLine.slice(0, 24);
+        const messagePart = logLine.slice(25);
+        const delimiter = logLine[24];
```

Comment:
Should probably add a a comment for these magic numbers since you documented `4`?

Reply:
Done!